rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs

```diff
@@ -97,90 +97,264 @@ impl From<(u64, Duration)> for Throughput {
     }
 }
 
-#[derive(Clone)]
-pub(super) struct ThroughputLogs {
-    max_length: usize,
-    inner: VecDeque<(SystemTime, u64)>,
-    bytes_processed: u64,
+/// Cell in a linear grid that represents a small chunk of time.
+#[derive(Copy, Clone, Debug)]
+enum Cell {
+    /// There is no data in this cell.
+    Empty,
+
+    /// No polling took place during this cell.
+    NoPolling,
+
+    /// This many bytes were transferred during this cell.
+    TransferredBytes(u64),
+
+    /// The user/remote was not providing/consuming data fast enough during this cell.
+    ///
+    /// The number is the number of bytes transferred, if this replaced TransferredBytes.
+    Pending(u64),
 }
+impl Cell {
+    fn merge(&mut self, other: Cell) {
+        use Cell::*;
+        // Assign values based on this priority order (highest priority higher up):
+        //   1. Pending
+        //   2. TransferredBytes
+        //   3. NoPolling
+        //   4. Empty
+        *self = match (other, *self) {
+            (Pending(other), this) => Pending(other + this.bytes()),
+            (TransferredBytes(other), this) => TransferredBytes(other + this.bytes()),
+            (other, NoPolling) => other,
+            (NoPolling, _) => panic!("can't merge NoPolling into cell"),
+            (other, Empty) => other,
+            (Empty, _) => panic!("can't merge Empty into cell"),
+        };
+    }
 
-impl ThroughputLogs {
-    pub(super) fn new(max_length: usize) -> Self {
+    /// Number of bytes transferred during this cell
+    fn bytes(&self) -> u64 {
+        match *self {
+            Cell::Empty | Cell::NoPolling => 0,
+            Cell::TransferredBytes(bytes) | Cell::Pending(bytes) => bytes,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+struct CellCounts {
+    /// Number of cells with no data.
+    empty: usize,
+    /// Number of "no polling" cells.
+    no_polling: usize,
+    /// Number of "bytes transferred" cells.
+    transferred: usize,
+    /// Number of "pending" cells.
+    pending: usize,
+}
+
+/// Underlying stack-allocated linear grid buffer for tracking
+/// throughput events for [`ThroughputLogs`].
+#[derive(Copy, Clone, Debug)]
+struct LogBuffer<const N: usize> {
```

Comment:
Is this our first usage of a GAT? If so, 🎊

Reply:
assuming you mean "const generic", I think yes!